File: domain/hanzi_candidate_types.py
"""Types for Hanzi candidate pipeline."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class HanziCandidate:
    """A single Hanzi candidate with provenance."""

    hanzi: str
    source: str  # e.g. "tier1", "tier2-char", "manual", "vocab", "cedict"
    freq: float = 0.0
    glosses: tuple[str, ...] = ()

    def with_glosses(self, glosses: Sequence[str]) -> "HanziCandidate":
        clean = tuple([g.strip() for g in glosses if isinstance(g, str) and g.strip()])
        return HanziCandidate(self.hanzi, self.source, self.freq, clean)

File: flutter_app/lib/domain/hanzi_candidate_types.dart
class HanziCandidate {
  final String hanzi;
  final String source;
  final double freq;
  final List<String> glosses;

  const HanziCandidate(
    this.hanzi,
    this.source,
    this.freq, [
    List<String>? glosses,
  ]) : glosses = glosses ?? const [];

  HanziCandidate withGlosses(List<String> glossesIn) {
    final clean = glossesIn
        .whereType<String>()
        .map((g) => g.trim())
        .where((g) => g.isNotEmpty)
        .toList();
    return HanziCandidate(hanzi, source, freq, clean);
  }
}
